src/hotspot/share/services/heapDumper.cpp

```diff
@@ -1857,21 +1876,25 @@ vframe* ThreadDumper::get_top_frame() const {
   return nullptr;
 }
 
-
-class VM_HeapDumper;
+// Callback to dump thread-related data for unmounted virtual threads;
+// implemented by VM_HeapDumper.
+class UnmountedVThreadDumper {
+ public:
+  virtual void dump_vthread(oop vt, AbstractDumpWriter* segment_writer) = 0;
+};
 
 // Support class using when iterating over the heap.
 class HeapObjectDumper : public ObjectClosure {
  private:
   AbstractDumpWriter* _writer;
   AbstractDumpWriter* writer()                  { return _writer; }
+  UnmountedVThreadDumper* _vthread_dumper;
 
   DumperClassCacheTable _class_cache;
 
  public:
-  HeapObjectDumper(AbstractDumpWriter* writer) {
-    _writer = writer;
-  }
+  HeapObjectDumper(AbstractDumpWriter* writer, UnmountedVThreadDumper* vthread_dumper)
```

Comment:
Maybe we can pass global writer to HeapObjectDumper and thus simplify the implementation? We can remove `class VM_HeapDumper : public VM_GC_Operation, public WorkerTask, public UnmountedVThreadDumper {` and `UnmountedVThreadDumper` then.

Reply:
This was an initial plan, but it would require more refactoring (dumper for unmounted vthreads needs a way to lock global writer, so need to move DumperController declaration and definition or add new 'class GlobalDumpWriter: public DumpWriter' and implement the lock there). I prefer to make functional changes here with minimum refactoring to simplify review and do refactoring in follow-up fixes